runtime/array.c

```diff
@@ -303,24 +306,81 @@ CAMLprim value caml_make_array(value init)
 
 /* Blitting */
 
+/* [wo_memmove] copies [nvals] values from [src] to [dst]. If there is a single
+   domain running, then we use [memmove]. Otherwise, we copy one word at a
+   time. Since the [memmove] implementation does not guarantee that the writes
+   are always word-sized, we explicitly perform word-sized writes to avoid
+   mixed-mode accesses.
+
+   [MM]: TODO enforcing the memory model also requires fences. See note in
+   memory.c */
+static void wo_memmove (value* const dst, const value* const src,
+                        mlsize_t nvals)
+{
+  mlsize_t i;
+
+  if (caml_domain_alone ()) {
+    memmove (dst, src, nvals * sizeof (value));
+  } else {
+    if (dst < src) {
+      /* copy ascending */
+      for (i = 0; i < nvals; i++) dst[i] = src[i];
+    } else {
+      /* copy descending */
+      for (i = nvals; i > 0; i--) dst[i-1] = src[i-1];
+    }
+  }
+}
+
 CAMLprim value caml_floatarray_blit(value a1, value ofs1, value a2, value ofs2,
                                     value n)
 {
-  memmove((double *)a2 + Long_val(ofs2),
-          (double *)a1 + Long_val(ofs1),
-          Long_val(n) * sizeof(double));
+  CAMLassert (sizeof(double) == sizeof(value));
+  wo_memmove(&Field(a2, Long_val(ofs2)), &Field(a1, Long_val(ofs1)),
+             Long_val(n));
```

Comment:
The address computations are wrong for 32 bit platforms!  Since we have not yet formally ruled them out, please use the same address calculations as in the current trunk, `(double *) a1 + Long_val(ofs1)`.  

More generally: please stop changing things irrelevant to multicore "because you can". it risks introducing new bugs, and it makes reviewing more painful than strictly necessary.